kratos/containers/pointer_vector_set.h

```diff
@@ -558,54 +558,108 @@ class PointerVectorSet final
-     * the sorting order and updates mSortedPartSize if necessary.
-     * @param Position An iterator pointing to the position where the pointer should be inserted.
-     * @param pData The pointer to be inserted.
+     /**
+     * @brief Inserts a pointer.
+     * @details This function inserts a given pointer such that the resulting PointerVectorSet
+     * is kept sorted. If there exists already a pointer with a key same as the key of the value, then
+     * this will return iterator of that existing pointer (The value will not be inserted.)
+     * @param value The pointer to be inserted.
      * @return An iterator pointing to the inserted element.
      */
-    iterator insert(iterator Position, const TPointerType pData)
-    {
-        ptr_iterator sorted_part_end;
-
-        key_type key = KeyOf(*pData);
-
-        if (mData.size() - mSortedPartSize >= mMaxBufferSize) {
-            Sort();
-            sorted_part_end = mData.end();
-        } else
-            sorted_part_end = mData.begin() + mSortedPartSize;
-
-        ptr_iterator i(std::lower_bound(mData.begin(), sorted_part_end, key, CompareKey()));
-        if (i == sorted_part_end) {
-            mSortedPartSize++;
-            return mData.insert(sorted_part_end, pData);
+    iterator insert(const TPointerType& value)
+    {
+        auto itr_pos = std::lower_bound(mData.begin(), mData.end(), KeyOf(*value), CompareKey());
+        if (itr_pos == mData.end()) {
+            // the position to insert is at the end.
+            mData.push_back(value);
+            return iterator(mData.end() - 1);
+        } else if (EqualKeyTo(KeyOf(*value))(*itr_pos)) {
+            // already found existing element with the same key, hence returning the existing element.
+            return iterator(itr_pos);
+        } else {
+            // insert the new value before the itr_pos.
+            return mData.insert(itr_pos, value);
         }
+    }
 
-        if (!EqualKeyTo(key)(*i))
-            if ((i = std::find_if(sorted_part_end, mData.end(), EqualKeyTo(key))) == mData.end()) {
-                mData.push_back(pData);
+    /**
+     * @brief Inserts a pointer at the specified position.
+     * @details This function inserts a given pointer. If the given position_hint is valid, then
+     * it uses that to insert the value, otherwise the position_hint is discarded to maintain the dataset
+     * sorted. If there is an existing element with the same key as in the value, then an iterator for
+     * the existing element is returned.
+     * @param position_hint An iterator pointing to the position where the pointer may be inserted.
+     * @param value The pointer to be inserted.
+     * @return An iterator pointing to the inserted element.
+     */
+    iterator insert(const_iterator position_hint, const TPointerType& value)
+    {
+        if (empty()) {
+            // the dataset is empty. So use push back.
+            mData.push_back(value);
```

Comment:
i would say you also need to update the sorted size somewhere

Reply:
What i was thinking, since we are going to be in a mixed state between `insert` and `push_back`, not to have the `mSortedPartSize` not initialized properly. So in this case, if someone uses `push_back` and `insert` in a mixed manner (unknowingly from different places), then still they will get the correct results.

Reply:
honestly this looks really dangerous to me...

Reply:
I can set the `mSortedPartSize = mData.size()` which should be correct if someone only uses the `insert`. would that be ok?

Reply:
yes

Reply:
done :)